### PR TITLE
Improve parallel scaling of MDNormSCD

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

A user noticed that the MDNormSCD algorithm runs much slower on our 12-32 core analysis computers than a 2-core laptop. A lot of time is spent in the critical section synchronizing updates to the signal array instead of doing work.

In this PR, I use `std::atomic::compare_exchange_weak` as suggested in [this reddit thread](https://www.reddit.com/r/cpp/comments/338pcj/atomic_addition_of_floats_using_compare_exchange/) to update the values of a temporary signal array before copying them into the MDHistoWorkspace. This appears to make a huge difference, reducing the execution time in the user's script from several minutes to 15 seconds (ndav3.sns.gov) and ~45 seconds (topaz.sns.gov)

**To test:**

@AndreiSavici Would you try this on one or more inputs and verify that it
  - Shows a considerable performance improvement
  - Still gives the correct result

@SimonHeybrock I would appreciate a second opinion on the atomic operations

There is no GitHub issue associated with this pull request.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by SimonHeybrock — COMMENTED** [review]
Had only a short look at this so far, will have to dig into `compare_exchange_weak` more to really verify this (but it looks reasonable, I think). Before I do that, two questions:

- Have you considered using, e.g., `std::atomic::fetch_add` (maybe with a custom memory order)?
- I do not really know how big `m_normWS->getNPoints()` is. If it is not too big, have you considered simply accumulating into a separate vector in each thread and adding them at the end? That would completely avoid any synchronization.

Apart from that, there is also some inefficiency due to the use of `SpectrumInfo::detector()`. If you can avoid that you might see some further speedup. I am also not sure that all the the `detid_t` to workspace index maps a required (but I did not check in detail).

**Comment by quantumsteve** [comment]
@SimonHeybrock 
- `std::atomic::fetch_add` is only available with integral and pointer types.
- The signal array is quite large (The user's example contained 701^3 doubles = 2.6 GB). Making 64 copies dramatically increases memory requirements.

File: Framework/MDAlgorithms/src/MDNormSCD.cpp [hunk 1]
```diff
@@ -363,6 +363,16 @@ void MDNormSCD::cacheDimensionXValues() {
   }
 }
 
+namespace {
+template <typename T, typename BinaryOp>
+void AtomicOp(std::atomic<T> &f, T d, BinaryOp _Op) {
+  T old = f.load();
+  T desired = _Op(old, d);
+  while (!f.compare_exchange_weak(old, desired))
+    desired = _Op(old, d);
```
Comment (SimonHeybrock) [inline Framework/MDAlgorithms/src/MDNormSCD.cpp]: Comparing this to the example in http://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange:
Why is the loop body needed?

Comment (quantumsteve) [inline Framework/MDAlgorithms/src/MDNormSCD.cpp]: It's possible that between lines 369 and 371 `f` was updated by another thread. The while loop updates the value of `desired` and tries again.

**Review by AndreiSavici — APPROVED** [review]
Works as advertised. On my computer I went from 26 seconds to 6. I checked the output workspaces, and they are the same.
:shipit:

File: Framework/MDAlgorithms/src/MDNormSCD.cpp (hunk 1 above)
Comment (SimonHeybrock) [inline Framework/MDAlgorithms/src/MDNormSCD.cpp]: Oh right, it is accumulation. :+1:

**Pull request merged**